### PR TITLE
refactoring ZK integration

### PR DESCRIPTION
**Pull request by kmrdhruv**

### CODE REVIEW

**Comment by gauravAshok** [comment]
We need to think more on the metastore requirements.

Entity Storage and their queriability is one thing. But in the interface we haven't provided any methods for distributed synchronization.

I was going through the Kafka's approach. https://github.com/apache/kafka/blob/trunk/metadata/src/main/java/org/apache/kafka/controller/Controller.java

During the zk based impl also, there was controller which I think was responsible for all mutations. That model is very much compatible to their new raft based state management. Basically Raft is replacing the zk's replicated log.

Here raft is just giving them the ability to choose controller (leader) among all broker nodes. 
https://github.com/apache/kafka/blob/trunk/raft/src/main/java/org/apache/kafka/raft/RaftClient.java

One that is done, all the delta are stored in topic and all the other brokers just maintain in-memory snapshot of metadata (all / partial? not sure yet).

Having a exclusive controller, means all transactions can be done in isolation without any need for synchonization.

So, what methods to we give to our metaStore api for this purpose? electLeader() / lock(path)?

Regarding the entity specific apis, I think we can go with targeted methods such as get/setTopic(), getAllTopics(), getController(), get/setSubscriptions() and so on, api set for each entity we want to be stored. 
(Kafka: in-mem metaCache: https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/MetadataCache.scala)

Need api methods for storing ephemeral entities? like registerApp(ip)?

**Comment by kmrdhruv** [comment]
I am thinking of not combining entity persistence with locking/leader election, but instead have a different abstractions for the same. Current implementation for later will be on ZK, but they both are different.  Locking/leader related abstraction will come later (i guess around when consumption path is implemented). 
So scope of current work is only metastore for Varadhi entities.
For entities, I didn't see variations in persistent APIs, primarily CURD, List and Search and hence common interface for all of the entities for now.

File: common/src/test/java/com/flipkart/varadhi/utils/JsonMapperTest.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+public class JsonMapperTest {
+
+    @Test
+    public void TestPolymorphicSerialization() {
+        Fleet myfleet = new Fleet();
+        myfleet.owner = "self";
+        myfleet.fleet.add(new Car("honda","classic"));
+        myfleet.fleet.add(new Bus(100,"Mahindra"));
+        String data = JsonMapper.jsonSerialize(myfleet);
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply common/src/test/java/com/flipkart/varadhi/utils/JsonMapperTest.java]: reverted the json changes.. will take them independently..

File: core/src/main/java/com/flipkart/varadhi/db/MetaStore.java [hunk 2]
```diff
@@ -1,12 +1,17 @@
...
+    // get(String parentKey, String resourceKey) and get(String resourceKey)
 
-    boolean exists(String resourceKey);
+    T get(String parent, String resourcePath);
+
+    T create(T entity);
+
+    boolean exists(String parent, String resourceName);
 
 }
```
[the remark this answers is not in the document]
Comment (kmrdhruv) [reply core/src/main/java/com/flipkart/varadhi/db/MetaStore.java]: more inclined towards higher level entities. No transactions as of now,  evolve later to have transactions or higher layer operations based on the need.

**Pull request merged**